# Аналитика в  книжном сервисе по подписке

<p style="font-size: 20px;">Описание проекта</p><br>
Коронавирус застал мир врасплох, изменив привычный порядок вещей. На какое-то время жители городов перестали выходить на улицу, посещать кафе и торговые центры. Зато стало больше времени для книг. Это заметили стартаперы — и бросились создавать приложения для тех, кто любит читать.
Ваша компания решила быть на волне и купила крупный сервис для чтения книг по подписке. Ваша первая задача как аналитика — проанализировать базу данных.
В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

<p style="font-size: 20px;">Цель проекта</p><br>
Изучить данные и выяснить какие книги или характеристики книг могут быть привлекательными для пользователей сервиса.

 <p style="font-size: 20px;">Описание данных</p>


Таблица books<br>
Содержит данные о книгах:

* `book_id` — идентификатор книги;
* `author_id` — идентификатор автора;
* `author_id` — идентификатор автора;
* `title` — название книги;
* `num_pages` — количество страниц;
* `publication_date` — дата публикации книги;
* `publisher_id` — идентификатор издателя.


Таблица authors<br>
Содержит данные об авторах:

* `author_id`— идентификатор автора;
* `author` — имя автораauthor;


Таблица publishers<br>
Содержит данные об издательствах:

* `publisher_id` — идентификатор издательства;
* `publisher` — название издательства;

Таблица ratings<br>
Содержит данные о пользовательских оценках книг:

* `rating_id` — идентификатор оценки;
* `book_id` — идентификатор книги;
* `username` — имя пользователя, оставившего оценку;
* `rating` — оценка книги.


Таблица reviews<br>
Содержит данные о пользовательских обзорах:
* `review_id` — идентификатор обзора;
* `book_id` — идентификатор книги;
* `username` — имя автора обзора;
* `text` — текст обзора.

 #### Импорт библиотек

In [1]:
import pandas as pd
from sqlalchemy import text, create_engine

#### Установка параметров и тестовый SQL запрос

In [2]:
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{user}:{pwd}@{host}:{port}/{db}'.format(**db_config)
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})
con=engine.connect()

#### Ознакомимся с таблицей books

In [3]:
# чтобы выполнить SQL-запрос, используем Pandas
query = '''SELECT * FROM books'''

pd.io.sql.read_sql(sql=text(query), con = con)

book_id  author_id                                              title  \
0          1        546                                       'Salem's Lot   
1          2        465                 1 000 Places to See Before You Die   
2          3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3          4         82  1491: New Revelations of the Americas Before C...   
4          5        125                                               1776   
..       ...        ...                                                ...   
995      996        571           Wyrd Sisters (Discworld  #6; Witches #2)   
996      997        454                        Xenocide (Ender's Saga  #3)   
997      998        201                                    Year of Wonders   
998      999         94                        You Suck (A Love Story  #2)   
999     1000        509  Zen and the Art of Motorcycle Maintenance: An ...   

     num_pages publication_date  publisher_id  
0          594       2005-11-01            93  
1          992       2003-05-22           336  
2          322       2010-12-21           135  
3          541       2006-10-10           309  
4          386       2006-07-04           268  
..         ...              ...           ...  
995        265       2001-02-06           147  
996        592       1996-07-15           297  
997        358       2002-04-30           212  
998        328       2007-01-16           331  
999        540       2006-04-25           143  

[1000 rows x 6 columns]

В таблице 1000 строк и 6 колонок

#### Ознакомимся с таблицей authors

In [4]:
query = '''SELECT * FROM authors'''

pd.io.sql.read_sql(sql=text(query), con = con)

author_id                                author
0            1                            A.S. Byatt
1            2        Aesop/Laura Harris/Laura Gibbs
2            3                       Agatha Christie
3            4                         Alan Brennert
4            5              Alan Moore/David   Lloyd
..         ...                                   ...
631        632         William Strunk Jr./E.B. White
632        633                           Zadie Smith
633        634                 Zilpha Keatley Snyder
634        635                    Zora Neale Hurston
635        636  Åsne Seierstad/Ingrid Christopherson

[636 rows x 2 columns]

В таблице 636 строк и 2 колонки

#### Ознакомимся с таблицей ratings

In [5]:
query = '''SELECT * FROM ratings'''

pd.io.sql.read_sql(sql=text(query), con = con)

rating_id  book_id        username  rating
0             1        1      ryanfranco       4
1             2        1   grantpatricia       2
2             3        1    brandtandrea       5
3             4        2        lorichen       3
4             5        2     mariokeller       2
...         ...      ...             ...     ...
6451       6452     1000  carolrodriguez       4
6452       6453     1000         wendy18       4
6453       6454     1000      jarvispaul       5
6454       6455     1000           zross       2
6455       6456     1000         fharris       5

[6456 rows x 4 columns]

В таблице 6456 рядов и 4 колонки

#### Ознакомимся с таблицей reviews

In [6]:
query = '''SELECT * FROM reviews'''

pd.io.sql.read_sql(sql=text(query), con = con)

review_id  book_id        username  \
0             1        1    brandtandrea   
1             2        1      ryanfranco   
2             3        2        lorichen   
3             4        3   johnsonamanda   
4             5        3     scotttamara   
...         ...      ...             ...   
2788       2789      999      martinadam   
2789       2790     1000         wknight   
2790       2791     1000  carolrodriguez   
2791       2792     1000         wendy18   
2792       2793     1000      jarvispaul   

                                                   text  
0     Mention society tell send professor analysis. ...  
1     Foot glass pretty audience hit themselves. Amo...  
2     Listen treat keep worry. Miss husband tax but ...  
3     Finally month interesting blue could nature cu...  
4     Nation purpose heavy give wait song will. List...  
...                                                 ...  
2788  Later hospital turn easy community. Fact same ...  
2789   Change lose answer close pressure. Spend so now.  
2790  Authority go who television entire hair guy po...  
2791  Or western offer wonder ask. More hear phone f...  
2792  Republican staff bit eat material measure plan...  

[2793 rows x 4 columns]

В таблице 2793 ряда и 4 колонки

#### Количество книг, вышедших после 1 января 2000 года

In [7]:
query = '''SELECT COUNT(*) AS num_books
               FROM books
               WHERE publication_date > '2000-01-01';'''

pd.io.sql.read_sql(sql=text(query), con = con)

num_books
0        819

Всего после 1 января 2000 года вышло 819 книг

In [8]:
# Проверим сколько книг вышло за предыдущие 20 лет
query = '''SELECT COUNT(*) AS num_books
               FROM books
               WHERE publication_date BETWEEN '1980-01-01' AND '2000-01-01';'''

pd.io.sql.read_sql(sql=text(query), con = con)

num_books
0        179

С 1980 по 2020 год вышло 179 книг.

#### Количество обзоров и средняя оценка для каждой книги

In [9]:
query = '''SELECT
            books.book_id,
            books.title,
            COUNT(DISTINCT reviews.review_id) AS review_count,
            AVG(ratings.rating) AS avg_rating
                FROM books
                LEFT JOIN reviews ON books.book_id = reviews.book_id
                LEFT JOIN ratings ON books.book_id = ratings.book_id
                GROUP BY books.book_id
                ORDER BY review_count desc, avg_rating desc;'''

pd.io.sql.read_sql(sql=text(query), con = con)

book_id                                              title  review_count  \
0        948                            Twilight (Twilight  #1)             7   
1        302  Harry Potter and the Prisoner of Azkaban (Harr...             6   
2        299  Harry Potter and the Chamber of Secrets (Harry...             6   
3        656                                     The Book Thief             6   
4        734                                   The Glass Castle             6   
..       ...                                                ...           ...   
995      191  Disney's Beauty and the Beast (A Little Golden...             0   
996      387                               Leonardo's Notebooks             0   
997      221                          Essential Tales and Poems             0   
998       83    Anne Rice's The Vampire Lestat: A Graphic Novel             0   
999      808                            The Natural Way to Draw             0   

     avg_rating  
0      3.662500  
1      4.414634  
2      4.287500  
3      4.264151  
4      4.206897  
..          ...  
995    4.000000  
996    4.000000  
997    4.000000  
998    3.666667  
999    3.000000  

[1000 rows x 4 columns]

Больше всего обзоров у Twilight - 7 обзоров, на втором и третьем месте серия книг Harry Potter, с 6 обзорами. Рейтинги варьируются от 1,5 до 5 баллов.

#### Издательство, выпустившее наибольшее число книг толще 50 страниц

In [10]:
query = '''SELECT
        publishers.publisher_id,
        publishers.publisher,
        COUNT(books.book_id) AS books_count
            FROM books
            JOIN publishers ON books.publisher_id = publishers.publisher_id
            WHERE books.num_pages > 50
            GROUP BY publishers.publisher_id
            ORDER BY books_count DESC
            LIMIT 1;'''

pd.io.sql.read_sql(sql=text(query), con = con)

publisher_id      publisher  books_count
0           212  Penguin Books           42

Издательством выпустившим больше всего книг (толще 50 страниц) оказалось Penguin Books, это издательство выпустило 43 книги.

#### Автор с самой высокой средней оценкой книг (только книги с 50 и более оценками)

In [11]:
query = '''SELECT
            authors.author_id,
            authors.author,
            AVG(filtered_books.avg_rating) AS avg_rating
        FROM
            (
                SELECT 
                    books.book_id, 
                    books.author_id,
                    AVG(ratings.rating) AS avg_rating
                FROM books
                JOIN ratings ON books.book_id = ratings.book_id
                GROUP BY books.book_id, books.author_id
                HAVING COUNT(ratings.rating_id) >= 50
            ) AS filtered_books
        JOIN authors ON filtered_books.author_id = authors.author_id
        GROUP BY authors.author_id, authors.author
        ORDER BY avg_rating DESC
        LIMIT 1;
        '''

pd.io.sql.read_sql(sql=text(query), con = con)

author_id                      author  avg_rating
0        236  J.K. Rowling/Mary GrandPré    4.283844

Автор с самым высоким средним рейтингом (при наличии 50 и более оценок) это Diana Gabaldon.

#### Среднее количество обзоров от пользователей, которые поставили больше 48 оценок

In [19]:
query = '''SELECT AVG(review_count) AS avg_reviews
        FROM (SELECT
                reviews.username,
                COUNT(reviews.review_id) AS review_count
            FROM reviews
            WHERE reviews.username IN (
                SELECT username
                FROM ratings
                GROUP BY username
                HAVING COUNT(rating_id) > 48
            )
            GROUP BY reviews.username
        ) AS user_reviews;
        '''

pd.io.sql.read_sql(sql=text(query), con = con)

avg_reviews
0         24.0

В среднем пользователи, поставившие более 48 оценок, оставили по 24 обзора на книги.

#### Вывод

<p style="font-size: 16px;">Всего после 1 января 2000 года вышло 819 книг:</p> <br>
Мы имеем данные по начало 2020 года. Что означает что за год выходило в среднем по 40 книг. Что примерно равняется 3 книгам в месяц. Это число может показывать что индустрия книг развивается за предушие 20 лет книг вышло почти в четыре раза меньше (хотя возможно в нашем приложении есть не все популярные книги с 1980 по 2020 год, данная иноформация требует уточнения)

<p style="font-size: 16px;">Больше всего обзоров у Twilight - 7 обзоров, на втором и третьем месте серия книг Harry Potter, с 6 обзорами. Рейтинги варьируются от 1,5 до 5 баллов.</p> <br>
По количеству обзоров мы можем судить о популярности конкретных книг. Например Twilight является популярной книгой не смотря на не высокий рейтинг, у нее большое количество поклонников и критиков. Данный срез по обзорам дает нам почву для анализа популярных книг, привлекающих интерес пользователей.

<p style="font-size: 16px;">Издательство, выпустившее наибольшее число книг толще 50 страниц</p> <br>
Издательство Penguin Books выпускает наибольшее количество книг на рынке. Возможно рассмотреть сотрудничество с данным издательством для получения уникальных условий для нашего сервиса, как продавца, и предложения издательству Penguin Books эксклюзивных условий по продвижению их книг.

<p style="font-size: 16px;">Автор с самой высокой средней оценкой книг (только книги с 50 и более оценками)</p> <br>
J.K. Rowling является автором с самыми высокими оценками среди читателей. Возможно выделение целевой аудитории автора в нашем сервисе, с целью предложения новых книг автора или книг аналогичной тематики. 

<p style="font-size: 16px;">Среднее количество обзоров от пользователей, которые поставили больше 48 оценок</p> <br>
Активные пользователи оставляют в среднем по 24 обзора. Обзоры пользователей и оценки произведений развивают сервис.Предлагаю рассмотреть стимулирование читателей оставлять отзывы, писать обзоры, получением скидок на книги или подписку. 